src/eos-updater-apply.c

```diff
@@ -73,41 +77,506 @@ get_test_osname (void)
+       * thrown if we succeeded at pulling the flatpak into the local
+       * repository but did not deploy it (since no FlatpakInstalledRef
+       * will be returned). */
+      flatpak_installation_install_full (installation,
+                                         FLATPAK_INSTALL_FLAGS_NO_DEPLOY,
+                                         flatpak_remote_ref_get_remote_name (to_install),
+                                         flatpak_ref_get_kind (FLATPAK_REF (to_install)),
+                                         flatpak_ref_get_name (FLATPAK_REF (to_install)),
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         NULL,
+                                         cancellable,
+                                         &local_error);
+
+      /* This is highly highly unlikely to happen and should usually only
+       * occurr in cases of deployment or programmer error,
+       * FLATPAK_INSTALL_FLAGS_NO_DEPLOY is documented to always return
+       * the error FLATPAK_ERROR_ONLY_PULLED */
+      if (!local_error)
+        {
+          g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "Flatpak installation should not have succeeded!");
+          cleanup_undeployed_flatpaks_in_installation (installation, flatpaks_to_deploy, i);
+          return FALSE;
+        }
+
+      /* Something unexpected failed, return early now.
+       *
+       * We are not able to meaningfully clean up here - the refs will remain
+       * in the flatpak ostree repo for the next time we want to install them
+       * but there's nothing in the public API that will allow us to get rid of
+       * them. */
+      if (!g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_ONLY_PULLED))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          cleanup_undeployed_flatpaks_in_installation (installation, flatpaks_to_deploy, i);
+          return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+static const gchar *
+get_pending_flatpak_deployments_state_path (void)
+{
+  return eos_updater_get_envvar_or ("EOS_UPDATER_TEST_UPDATER_FLATPAK_UPGRADE_STATE_DIR",
+                                    LOCALSTATEDIR "/lib/eos-application-tools");
+}
+
+/* We return a transfer-full GFile of the record of flatpaks to be
+ * deployed on the next boot. If something goes wrong the caller is expected
+ * to delete the file. In any event, the caller must free the return value. */
+static GFile *
+write_flatpaks_to_be_deployed (GPtrArray    *flatpaks_to_deploy,
+                               const gchar  *deployment_csum,
+                               GCancellable *cancellable,
+                               GError       **error)
+{
+  const gchar *pending_flatpak_deployments_state_path = get_pending_flatpak_deployments_state_path ();
```

Comment:
Since this variable is used only in the `g_build_filename()` call below, consider calling the getter there directly to reduce the number of variables a bit